File: 13week/LeetCode-104-Maximum-Depth-of-Binary-Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxDepth(self, root: list[TreeNode]) -> int:
        def depth(node):
            if node is None:
                return 0
            return max(depth(node.left), depth(node.right))+1
        
        return depth(root)

File: Tree/Leetcode-104-MaximumDepthOfBinaryTree/Leetcode-104-MaximumDepthOfBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution { 
public:
    int result = 0;
    int maxDepth(TreeNode* root, int depth = 1) {
        if(root == nullptr) return 0;
        
        result = max(result, depth);
        if(root->left){
            maxDepth(root->left, depth + 1);
        }
        if(root->right){
            maxDepth(root->right, depth + 1);
        }
        return result;
    }
};